man/man1/cft2clm.1
.TH cft2clm 1  "USER COMMANDS"
.SH NAME
cft2clm \- convert binary land use data files to clm data files for LPJmL
.SH SYNOPSIS
.B cft2clm
[\-h] [\-firstyear first] [\-lastyear last] [\-nyear n] [\-nbands n] [\-swap]
.I cftfile clmfile
.SH DESCRIPTION
Program converts a raw binary land use data file consisting of 16bit integer values into a file format suitable for LPJmL. It is an alias to
.B addheader
-nbands 64 -firstyear 1700 -lastyear 2005 -scale 0.001 -ncell 0 -id LPJLUSE.
.SH OPTIONS
.TP
\-h
Print a short help text.
.TP
\-firstyear first
First year in land use file (default is 1700).
.TP
\-lastyear  last
Last year in land use file (default is 2005).
.TP
\-nyear n
Number of years in land use file (default is 306).
.TP
\-nbands n
Number of bands in output file (default is 64).
.TP
\-swap
Change byte order in land use file
.TP
.I cftfile
Filename of binary land use data file.
.TP
.I clmfile
Filename of clm data file written.
.SH EXAMPLES
.TP
Create clm data file from land use dataset changing byte order:
.B cft2clm
\-swap cft64.bin cft_1500-2017.clm
.PP
.SH EXIT STATUS
.B cft2clm
returns a zero exit status if the clm file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjml(1), addheader(1), cru2clm(1), txt2clm(1), grid2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: David Hötten (davidho@pik-potsdam.de)

### Added

- Warning message added in `lpjml`, `lpjcheck`,`cdf2soil`, `txt2grid` and `cdf2coord` if cell size cannot be resolved by grid data in short format.
- Function `isfloatcoord()` added in `coord.c` to check whether coordinate must be stored as a float variable, test function `test_isfloatcoord.c` added.
- Test functions for `vmaxlimit_tree`, `vmaxlimit_grass`, `vmaxlimit_crop`, `bisect`, `leftmostzero` and `getbufferavg` functions added.

### Changed

- Error about first simulation year is less than first year of climate data replaced by warning in `initclimate.c`.
- First and last simulation year in spinup changed from 1901 to 1840 in `lpjml_config.cjson`.
- Number of spinup years changed from 3500 to 3439 in `lpjml_config.cjson` in order to get bit-identical results.
- Test files for unit test added to tar and zip file in `Makefile`.

### Removed

- Unused source file `src/lpj/part.c` removed.

### Fixed

- Expected number of bands set correctly for raw input files in `openinputdata.c`.
- Hard-coded filename `test.lpj` replaced by temporary filename in `test_append.c`, `test_getnoread.c`, `test_indexarray.c`, `test_isdefined.c`, `test_isnull.c`, `test_restart.c`, and `test_writename.c` in order to allow parallel execution of tests.
- Arguments corrected in calls of lpjml functions in `lpj_poem.c`.
- Missing `-print_noread` option added in `INSTALL`.
- `setupannual_grid.c` corrected to compile with `-DIMAGE -DCOUPLED` flag.
- Error message for different datatypes corrected in `lpcat.c`.
- Variables initialized to zero and NULL in `phenology_gsi.c`, `lpjcat.c` and `restart2yaml.c` to avoid warnings in `gcc`.
- Man page for `json2restart` corrected.
- Missing option `-nooutput` added to man page of `lpjfiles`.
- Help text for `-h` option of `printclm` corrected.
- Check for NULL pointer for `pfts` array corrected in `lpjcat.c`.
- Double entry of `getlag.$O` removed in `src/soil/Makefile`.


## [5.10.1] - 2025-09-30

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review: Maik Billing (billing@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- Variable `phen` added to output in `fprintpft.c`.
- Compile option `NRECOVERY_COST` added to `INSTALL` and man page of `configure.sh`.
- Check for valid `swc_bnf` parameters added in `fscanpftpar.c`.
- Wirth et al. (2024)  article added to `REFERENCES`.
- New parameter `"tscal_b"` added in `lpjparam.cjson`.
- New virtual function `getb()` added to type `Pft`.

### Changed

- 2 boolean flags `phen_to_one` and `rainyseason` added to tree parameter. These are used in `phenology_gsi.c` instead of string compare with tree names.
- Unused argument `aet_layer` removed from `nitrogen_stress()` function.
- `phi_bnf` parameter calculated from `swc_bnf` parameter instead of reading from `pft.cjson` file in `fscanpftpar.c`.
- Hard-coded parameter in temperature dependent calculation of `b` replaced by global parameter `tscal_b`.
- Calculation of `b` in `setupannual_grid.c` replaced by call to new virtual PFT function `getb()`.

### Removed

- Parameter `phi_bnf` removed from `pft.cjson`.
- Variable `b` removed from restart file.
- Unused macro `RESTART_VERSION` removed from `header.h`.

### Fixed

- Arguments of `npp` function corrected in `daily_woodplantation.c`. The current version leads to a too strong reduction in NPP.
- `nuptake_crop.c` changed to compile with `-DDEBUG_N` option.
- `update_daily_cell.c` changed to compile with `-DDEBUG` option.
- `daily_agriculture_tree.c` changed to compile with `-DDEBUG2` option.
- Typo in error message in `bstruct_writearrayindex.c` corrected.
- Error message corrected in `bstruct_wopen.c`.
- Error number corrected in `bstruct_finish.c`.
- Argument `temp` added to `f_turnover_tree` function to compile with `-DNRECOVERY_COST` option.


## [5.10.0] - 2025-08-14

### Contributors

- author: Jens Heinke (heinke@pik-potsdam.de), Sibyll Schaphoff (sibylls@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- option to specify year after which deforestation is no longer feeding product pools but biomass is burnt instead: `luc_burn_startyear` (disabled)
- option to specify costs for N recovery upon turnover (pre-compiler switch `NRECOVERY_COST`, disabled)

### Changed

- parameter changes:
  - `alphaa` from 0.5 to 0.7 for natural vegetation;
  - `theta` from 0.9 to 0.95;
  - `fn_turnover` PFT1: 0.5, PFT2: 0.5, PFT3: 0.7, PFT4: 0.3, PFT5: 0.3, PFT6: 0.7, PFT7: 0.3, PFT8: 0.3; updated values from [Sophia et al. 2024](https://doi.org/10.5194/bg-21-4169-2024) for all other PFTs except crops
  - `cnratio_leaf.low/median/high` for all PFTs updated from TRY data
  - `b` (ratio of dark respiration Rd to maximum photosynthetic capacity Vcmax) set to 0.031 for C3 and C4 plants (value for 25°C, from [Wang et al. 2020](https://doi.org/10.1111/gcb.14980), consistent with TRY data)

src/utils/cft2clm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                  c  f  t  2  c  l  m  .  c                                     \n**/
/**                                                                                \n**/
/**     Program converts cftfile to LPJ *.clm file                                 \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include <sys/stat.h>
#include "lpj.h"

#define NBAND 26
#define NYEAR 306
#define FIRSTYEAR 1700 

#define USAGE "Usage: %s [-h] [-swap] [-nbands n] [-firstyear s] [-nyear y] [-lastyear e] cftfile outfile\n"

int main(int argc,char **argv)
{
  FILE *ifile,*ofile;
  short *data;
  int i,n,year,count,band;
  Header header;
  Bool swap;
  struct stat filestat;
  char *endptr;
  swap=FALSE;
  header.nyear=NYEAR;
  header.firstyear=FIRSTYEAR;
  header.order=CELLYEAR;
  header.firstcell=0;
  header.nbands=NBAND;
  header.cellsize_lon=header.cellsize_lat=0.5;
  header.datatype=LPJ_SHORT;
  header.scalar=0.001;
  for(i=1;i<argc;i++)
    if(argv[i][0]=='-')
    {
      if(!strcmp(argv[i],"-h"))
      {
        printf("%s (" __DATE__ ") - adds header to cftfile for LPJmL version %s\n\n",argv[0],getversion());
        printf(USAGE
               "\nArguments:\n"
               "-h           print this help text\n" 
               "-swap        change byte order in cftfile\n"
               "-nbands n    number of bands, default is %d\n"
               "-firstyear s first year, default is %d\n"
               "-lastyear e  last year, default is %d\n"
               "-nyear y     number of years, default is %d\n"
               "cftfile      filename of input cft data file\n"
               "outfile      filename of output cft data file\n\n"
               "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n",
               argv[0],NBAND,FIRSTYEAR,FIRSTYEAR+NYEAR-1,NYEAR);
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[i],"-swap"))
        swap=TRUE;
      else if(!strcmp(argv[i],"-nbands"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Missing argument after option '-nbands'.\n"
                  USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        header.nbands=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-nbands'.\n",argv[i]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-nyear"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Missing argument after option '-nyear'.\n"
                  USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        header.nyear=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-nyear'.\n",argv[i]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-firstyear"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Missing argument after option '-firstyear'.\n"
                  USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        header.firstyear=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-firstyear'.\n",argv[i]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-lastyear"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Missing argument after option '-lastyear'.\n"
                  USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        header.nyear=strtol(argv[++i],&endptr,10)-header.firstyear+1;
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-lastyear'.\n",argv[i]);
          return EXIT_FAILURE;
        }
      }
      else
      {
        fprintf(stderr,"Error: invalid option '%s'.\n",argv[i]);
        fprintf(stderr,USAGE,argv[0]);
        return EXIT_FAILURE;
      }
    }
    else 
      break;
  if(argc<i+2)
  {
    fprintf(stderr,"Filenames missing.\n");
    fprintf(stderr,USAGE,argv[0]);
    return EXIT_FAILURE;
  }
  ifile=fopen(argv[i],"rb");
  if(ifile==NULL)
  {
    fprintf(stderr,"Error opening '%s': %s\n",argv[i],strerror(errno));
    return EXIT_FAILURE;
  }
  fstat(fileno(ifile),&filestat);
  n=filestat.st_size/sizeof(short)/header.nyear;
  if(filestat.st_size % (sizeof(short)*header.nyear*header.nbands))
    fprintf(stderr,"Warning: File size is not multiple of nbands * nyear\n"); 
  printf("Number of cells: %d\n",n/header.nbands);
  ofile=fopen(argv[i+1],"wb");
  if(ofile==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s\n",argv[i+1],strerror(errno));
    return EXIT_FAILURE;
  }
  header.ncell=n/header.nbands;
  fwriteheader(ofile,&header,LPJ_LANDUSE_HEADER,LPJ_LANDUSE_VERSION);

  data=(short *)malloc(n*sizeof(short));
  if(data==NULL)
  {
    printallocerr("data");
    return EXIT_FAILURE;
  }
  for(year=0;year<header.nyear;year++)
  {
    if(freadshort(data,n,swap,ifile)!=n)
    {
      fprintf(stderr,"Unexpected end of file in year %d.\n",year+header.firstyear);
      return EXIT_FAILURE;
    }
    count=0;
    for(band=0;band<n;band++)
      if(data[band]>0)
        count++;

    printf("year:%d count:%d\n",year+header.firstyear,count);

    if(fwrite(data,sizeof(short),n,ofile)!=n)
    {
      fprintf(stderr,"Error writing data in '%s': %s.\n",argv[i+1],strerror(errno));
      return EXIT_FAILURE;
    }
  }
  fclose(ifile);
  free(data);
  fclose(ofile);
  return EXIT_SUCCESS;
} /* of 'main' */
